python/unitytrainers/self_play_trainer_controller.py

```diff
@@ -0,0 +1,64 @@
+import os
+import tensorflow as tf
+from unitytrainers import TrainerController, UnityTrainerControllerException
+
+
+class SelfPlayTrainerController(TrainerController):
+
+    def __init__(self, env_path, run_id, save_freq, curriculum_folder, fast_simulation, load, train,
+                 worker_id, keep_checkpoints, lesson, seed, docker_target_name, trainer_config_path,
+                 no_graphics):
+        super(SelfPlayTrainerController, self).__init__(self, env_path, run_id, save_freq,
+                                                        curriculum_folder, fast_simulation,
+                                                        load, train, worker_id, keep_checkpoints,
+                                                        lesson, seed, docker_target_name, trainer_config_path,
+                                                        no_graphics)
+        self.ghost_trainers_path = self.model_path + 'ghost/'
+        if not os.path.exists(self.ghost_trainers_path):
+            os.makedirs(self.ghost_trainers_path)
+
+        # TODO add opci, delta and opponent save frequency
+
+    def _initialize_trainers(self, trainer_config, sess):
```

Comment:
I'm not sure where this and `initialize_ghost_trainer_savers(self, ghost_trainers)` are called. Should they be called in the constructor?

Reply:
the function `_initializer_trainers` is an overriden function, the original one coming from `trainer_controller`. It is called in the `initialize_training_environment`, which in turn is a function that we refactored (previously part of `start_learning`). It should not be part of the constructor, as it takes a session as an argument, and the session isn't created until inside `start_learning`.

Reply:
ok sounds good to me. The only thing then is, why does it start with `_`

Reply:
Because Python is too cool to care about encapsulation, a method is considered "private" if it starts with an underscore. I personally dislike this convention. It starts with underscore because the original method in the `TrainerController` starts with an underscore.